### DISCUSSION TITLE
Using the internal temperature option, but to prevent heating when over target.

### DISCUSSION THREAD

**Discussion by t0mb080** [jmcollin78/versatile_thermostat#569] [General]
Firstly, thank you for making such an excellent thermostat integration.  I have aircon units but the internal temperature control is terrible, so your integration is extremely useful.

I have checked the option to use the internal temperature of the units to 'force' heating when it is required which solves the problem of the internal temperature sensor heating up and turning off the heat even though room temperature is below target.  However, the internal control of the unit is so bad that I also have the same issue in the other direction, i.e. when not calling for heat.  For example, the target temperature is 20, the room temperature is 21  so VTherm sends a target to the unit of 20 or similar but the unit starts heating anyway!  Is there a setting to somehow fix this, i.e. send a very low temperature when no heating is required to prevent erroneous heating?

Maybe one way to do it would be to have an option to treat the control over climate more like a switch and just specify a very high target number (e.g. 30) when VTherm detects heat is needed and another very low target (e.g. 6) when heat is not needed just to guarantee the unit is in the right state?

**Comment by t0mb080** [comment]
An alternative scheme, which I think would be really useful for AC units with extremely poor self-regulation would be something like the logic below.  Would you consider adding a mode like this?  Or maybe adding parameters to the expert regulation such that this could be achieved with the right parameter choice?

Thanks again for a great extension.

```
def calculate_heating_action(current_room_temperature, target_room_temperature, is_heating_state, offset=5, threshold=0.25):
    '''
    A simple controller which forces the climate entity into a heat or not heat state using an offset around the heating target.
    Used for climate entities with very poor self-regulation to reliably force them into a heating or non-heating state.

    The threshold parameter defines a band around the target temperature such that it only starts or stops heating if it's moved
    outside the band to prevent constant switching on or off.
    '''

    if is_heating_state:
        # in heating state, check whether we should turn off
        if current_room_temperature > (target_room_temperature + threshold):
            is_heating_state = False
            climate_entity_target = current_room_temperature -  offset
    else:
        # not in heating state, check whether  we should turn on
        if current_room_temperature < (target_room_temperature - threshold):
            is_heating_state = True
            climate_entity_target = current_room_temperature + offset

    return is_heating_state, climate_entity_target

```

**Comment by t0mb080** [comment]
You can see the issue graphically below.  See that the VTherm target in yellow does a good job of ramping up high to force the heating on, which then drags the room temperature (red) up to the desired temperature (green).  However, see what happens at around 16:50, the desired temperature is reached so VTherm reduces the target to the underlying in yellow but because of the poor regulation of the underlying AC system it's not low enough because it fires up again (see blue line which is temperature of the actual AC) even though it's over target.

![image](https://github.com/user-attachments/assets/5e6f2a5d-376c-4f6a-b702-bc84821ef8ba)

**Reply by jmcollin78** [reply]
Hello @t0mb080 ,

Your graph is very interesting and shows the job done by VTherm self-regulation with 'use internal temperature' checked.
What I see here around 16:50 is pretty good, the temperature is decreasing, very near the target, so the heater starts to heat but for a very little time (one minute ?).

![Capture d’écran 2024-10-26 à 10 54 57](https://github.com/user-attachments/assets/dc03c6af-794d-47d0-9073-62a6cce96452)

Regulation and self-regulation should be look over a long period to see if it converge or not. Sadly we do not see in your graph what happens after 16:54, but I suppose we will have small variation around the target.

For me this is what is expected and you will not have better result with a pure On/Off algorithm as you propose.

If you want to experiment it, you can always create a virtual switch that implements your logic (there is example in the readme) and create a `over_switch` Vtherm with this virtual switch.

**Comment by t0mb080** [comment]
Thanks very much for the reply @jmcollin78.

You are right in that the self-regulation generally does converge after it's up to temperature, but it can take quite a while because the underlying continually fires up every ten minutes or so which sends the temperature above target but eventually the integrated error accumulates high enough to force the target low enough to keep the heating off.  I suppose my fundamental question is why is the "use internal temperature" option asymmetric, i.e. why does it act to force heating but not act in the opposite way to prevent heating?  The internal regulation of my underlying is equally bad in both directions sadly!

I did consider specifying it as a switch as you suggest (after all it really is a binary device, i.e. so-called bang-bang control in essence, just through the added complication of the temperature targets) but I couldn't work out how to achieve the other features you get with the climate control - most importantly getting the units to actually turn on and off when needed, either manually or by scheduler.

**Reply by jmcollin78** [reply]
> why does it act to force heating but not act in the opposite way to prevent heating?

It works in both way.
If the current_temp is 20, internal_temp is 15 (-5 delta) and target is 19, then the regulated_temp will be 20 - (20-19) - 5 -> 14. And the device shouldn't heat (14 vs 15). The compensation is always applied and so works in both way

**Reply by t0mb080** [reply]
Ah, I see.  In that case, I suppose the resulting 1 degree offset just isn't severe enough to force my units into the correct state.  To use your example of a 15 internal_temp and a 14 regulated_temp, I completely agree that the device shouldn't heat under these conditions but sadly mine actually frequently does in this scenario!  I don't know why, maybe there's some additional logic in there, or a hidden temperature offset that's not viewable externally.   I'll keep playing with it to see whether there's some configuration I can use to really force it into the state the VTherm thinks it should be in.

**Comment by t0mb080** [comment]
Here's an example of the behaviour I mean.   Here you can see that the room temperature (red) goes over target (green), with the internal temperature (blue) showing that the unit is still periodically firing out heat even though the regulated temperature (yellow) from VTherm is below target.  You can see that over a period of three hours or so, the integrated error eventually drives the regulated temperature (yellow) low enough that it forces the heating off so the self regulation does work, it's just it would work much more efficiently if it sent a lower target to force the heating off sooner so it didn't rely on the error integration to force the target low enough.

![image](https://github.com/user-attachments/assets/5ffb7a0b-0a60-4d73-b32a-87fc8e36bb8f)

**Reply by jmcollin78** [reply]
Have the strong regulation activated ?
If no, try it, you are clearly in the case of the strong regulation need + use internal temp.

If yes, you should maybe try much higher regulation to go down more quickly. There is a chapter on this subject on the README documentation: https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#self-regulation-in-expert-mode

You can try this values:

kp: 0.5
ki: 0.08
k_ext: 0.0
offset_max: 8
stabilization_threshold: 0.1
accumulated_error_threshold: 80

**Reply by t0mb080** [reply]
Thanks very much @jmcollin78.  I am using the strong regulation but I'll give the expert mode a try and look at the values you suggest.

**Comment by t0mb080** [comment]
I'm wondering if something is actually wrong.  See the plot below where I've used the same colour scheme as previously but I've added the regulated temperature as provided by VTherm in purple.  I was expecting this to match the yellow which is the actual target reported by the underlying.  Have I misunderstood or is something actually wrong here?

![image](https://github.com/user-attachments/assets/f2ba76ee-1fc9-49da-9f54-10d0d6a57c3a)

**Reply by t0mb080** [reply]
The purple regulated temperature looks very sensible, falling as the room temperature (red) goes over target.  However, the link between that and the actual target that gets fed to the underlying (yellow) isn't at all obvious, and the system is firing up the heating from 1710 to 1800 even though the regulated temperature is much lower than target throughout this period.  Is the offset being applied correctly here?  I notice the blue (underlying internal temperature) updates and changes much more frequently than the regulated and target being sent to the underlying (yellow), is this the cause of the mismatch?

**Comment by t0mb080** [comment]
Here's another example of the same thing.  The regulated temperature (purple) looks very sensible but it doesn't translate through to the actual target sent to the underlying (yellow) because the internal temperature offset just can't cope with the fast swings in the underlying temperature so it ends up still pumping out heat and overshooting even it's well above target and the regulated temperature well below.

![image](https://github.com/user-attachments/assets/343e02e0-1fe2-4268-bb6d-87ecf3662190)

Possible solutions I can think of are:

1)  Increase the sample rate of the adjustment to keep up with the underlying.  I don't think this is a good idea as it would feed out very fast changes in set temperature which may be an issue for the physical unit, and also possibly for the API the underlying could be using.

2)  Have an option to make the internal offset option asymmetric.  This might work as it seems very effective at forcing heating on but much less effective at getting the heating off and preventing the kind of overshoot seen in my plots.  All the implications might be hard to predict though and it might have some weird behaviour where the target and actual are close so it flips states?

3)  Play further with the expert regulation parameters?  As I understand it, this probably wouldn't work as it would affect the regulated temperature (the purple line in my plots).  However, this is already very good, the real issue is the translation of that line to the actual target fed to the underlying through the internal termperature offset mechanism.

4)  Add an option like my proposal in the second post.  I realise this is a very basic control scheme but it would also be extremely robust and as long as the offset was set high enough it would work even with large swings in the underlying temperature like I've shown with my units.  It could be a great fallback option for units with problems like mine - which I don't think are that uncommon, these are standard Panasonic AC wall mounted units with (unfortunately) the internal temperature sensor in the unit itself which causes it to swing around wildly as it heats.

I'd love your input @jmcollin78 and if it helps convince you that (4) would be a worthwhile addition to your great integration then I'd be very willing to make a decent donation!

**Comment by jmcollin78** [comment]
Hello is something strange in your graph.
![Capture d’écran 2024-10-31 à 10 21 10](https://github.com/user-attachments/assets/58ab09a8-cfc8-47a2-a6ec-ea7740b8b0c6)

Why is the TV room target in yellow (I think it is the internal target of the device) not following the regulated temperature (in purple) ?
This should be the case, because the regulated temperature is send to device as target temperature. There is something not working here which explains the overheating period (cf. screenshot)

**Reply by t0mb080** [reply]
I assumed that the regulated temperature was the target before the "Use internal device temperature" option offset was applied?  This certainly looks to be the case from the data, as the yellow does appear to have been boosted towards the internal device temperature (blue).   The regulated temperature certainly doesn't appear to have the offset added, as it's nowhere near the blue?  Is this not the case though?

**Reply by t0mb080** [reply]
Also, confirming this, if I turn the option off to use the internal device temperature, the yellow immediately jumps to match the purple.  Surely this proves it?

**Reply by jmcollin78** [reply]
You are right. The regulated temp doesn't include the compensation of temperature (internal temp - room temp).

Your proposal will work in only very specific case, and I try to avoid too much specific case, because there is one specific case per user (almost). There is already so many options that users are lost, and I really care to keep it simple as much as possible. you can see in the issues or discussion, the time I spent to explain the options.

Moreover, I'm pretty sure, the solution is already there.

> As I understand it, this probably wouldn't work as it would affect the regulated temperature (the purple line in my plots)

In your case, and because the internal temperature is finally not reliable, **I would try to switch off** the 'use internal device temperature' and use the expert regulation mode. With the offset max of 8, the regulated temp should climb up to the very bad internal temp of your device.

Do you give it a try as suggested above ?

**Reply by t0mb080** [reply]
Thanks @jmcollin78, I understand your desire not to complicate things although I think my suggestion would work in almost all cases even if it wouldn't be the best option in most of them, I would view it as a fallback option if the more sophisticated options don't work.  

I had the same thought as you about turning the internal temperature correction off and relying on expert regulation to turn up the regulation high enough to force control of the underlying.  I'm very willing to try this out although I have a slight practical issue in that I've added the parameters to my configuration.yaml but when I turn on expert mode, the thermostat immediately snaps the regulated termperature straight to the target. I assume this is because it wasn't able to read the configuration properly but I haven't been able to work out how to fix it yet.

**Reply by jmcollin78** [reply]
Do you have errors in logs ?

**Reply by t0mb080** [reply]
Apologies @jmcollin78 , I haven't check the logs - I'll check the documentation on how to do that and check.

Actually, I do have a compromise suggestion of how I could achieve what I want with the addition of a single parameter to your expert regulation.  All that would be needed is to add the 'threshold' parameter in my original scheme to your expert regulation such that you simply check if the room temperature is within the target +/- threshold and if so you simply don't do anything further so the regulated temperature doesn't move.  This would have the following advantages:

1)  It wouldn't materially complicate your great integration, it's only the addition of a single very useful parameter to the expert regulation and if people are using the expert regulation then they're likely a tricky or technical case anyway and so could benefit from such a useful and interpretable parameter.

2) It would default to zero so there would be no change in the default behaviour.

3) It would enable people with units like mine to control them well because it would enable us to turn up the kp parameter very high to force control of the unit to completely override the internal temperature whilst not suffering the current downside of this of magnifying tiny deviations around the target and flip-flopping around zero.

**Reply by jmcollin78** [reply]
https://github.com/jmcollin78/versatile_thermostat/tree/6.5.0.beta1?tab=readme-ov-file#self-regulation-in-expert-mode

**Reply by t0mb080** [reply]
Thanks @jmcollin78.  I've copied the parameters exactly as specified in the docs into my configuration.yaml.  I've looked in the logs and get the following:

VersatileThermostat-TV room thermostat - Cannot initialize Expert self-regulation mode due to VTherm API doesn't exists. Please contact the publisher of the integration

**Reply by t0mb080** [reply]
This is the exact contents of my configuration.yaml:

```

# Loads default set of integrations. Do not remove.
default_config:

# Load frontend themes from the themes folder
frontend:
  themes: !include_dir_merge_named themes

automation: !include automations.yaml
script: !include scripts.yaml
scene: !include scenes.yaml


versatile_thermostat:
     auto_regulation_expert:
         kp: 0.4
         ki: 0.08
         k_ext: 0.0
         offset_max: 5
         stabilization_threshold: 0.1
         accumulated_error_threshold: 50

```

**Reply by t0mb080** [reply]
I have verified that the configuration file is definitely getting read, because if I put some rubbish in it, home assistant complains.  Home assistant doesn't complain about that VTherm expert config though, so it is valid YAML at least.

**Reply by jmcollin78** [reply]
Yes it seems good. Do you have activated the expert mode regulation ? (In VTherm configuration)

I'm looking the code on my side.

Here is my I have on my side:
configuration.yaml:
```
versatile_thermostat: !include versatile_thermostat.yaml
```
versatile_thermostat.yaml:
```
auto_regulation_expert:
  kp: 0.4
  ki: 0.08
  k_ext: 0.0
  offset_max: 4
  stabilization_threshold: 0.1
  accumulated_error_threshold: 40
safety_mode:
  check_outdoor_sensor: false
```

try to put only 2 blank for indentation. I see you have 6 blanks. Maybe it is the issue

**Reply by t0mb080** [reply]
Ah, I have, but only in one thermostat - do I have to switch them all to expert mode for it to work?

**Reply by jmcollin78** [reply]
All VTherm that you want to be in expert mode for sure.  Not more.

Do you have seen this: `try to put only 2 blank for indentation in your configuration.yaml. I see you have 6 blanks. Maybe it is the issue`

**Reply by t0mb080** [reply]
It wasn't the spaces.  However, I restarted the whole home assistant and it magically started working.  

Thanks very much, I'll play with the expert parameters and report back.

**Reply by jmcollin78** [reply]
Thank you for the beer ! I just receive them. It is very pleasant 👍

**Comment by t0mb080** [comment]
Ok, I've done some initial playing with some fairly encouraging and interesting results.  One thing to note, is that in order for VTherm to correctly update the expert parameters in the configuration.yaml, it seems to require a complete restart of Home Assistant - neither reloading the yaml, nor reloading the VTherm integration seems to do it.  I don't know whether this is expected, or is specific to my setup - I'm running in docker if that makes any difference to anything.

Anyway, I've turned the internal temperature compensation off as you suggested because it's clear that it can't cope with the big and frequent swings in my units internal temperature which was causing the overshoot we saw previously.  This means that I have to rely on very big gain in the self-regulation to force my units to reliably either heat or not heat to overcome the big swings in internal temperature.  My initial experiment at this has been to simply set the kp value to 5.0(!).  This actually works quite well as you see below.  When the heating comes on at 8am, the error is big enough to basically force the unit to constantly heat to quickly get to target by about 0830.  It then keeps it quite well at target, although it does involve quite big swings in the regulated temperature to do so and with some oscillation.

This is much better than it was as it is able to quickly get to target and not overshoot too much.  I do still think adding a 'threshold' parameter to the expert regulation such that it doesn't update the regulated temperature when it's within +/-threshold of target might be a valuable addition as it would enable high gains without swinging around when close to target which would reduce the frequency of the heating cycle and therefore maybe increase efficiency and reduce wear on the heating units.

![image](https://github.com/user-attachments/assets/fb1642c7-687c-4cb6-a1a5-8aeddd2e4e77)

**Reply by jmcollin78** [reply]
> neither reloading the yaml, nor reloading the VTherm integration seems to do it

You have an option in Development Tools / Yaml to reload all VTherm configuration. This should reload also the configuration.yaml:
![Capture d’écran 2024-11-01 à 11 42 30](https://github.com/user-attachments/assets/2a63da03-9796-4bc8-b901-0c4ca9b65289)

You have a very stable and nice regulation curve !. Small oscillations around target, stable current temperature, ... really fine.
May I take it as an example in my README file ?

> I do still think adding a 'threshold' parameter to the expert regulation such that it doesn't update the regulated temperature

There is already thresholds in expert mode: stabilization_threshold and accumulated_error_threshold
Have you try to change the `stabilization_threshold` which was designed for what you need (if I understand well, but I'm not totally sure).

**Reply by t0mb080** [reply]
Thanks @jmcollin78, yes, of course you may use it as an example.  I think it's a good example to have, because it shows where an alternative approach to the internal temperature option is needed where the internal device regulation is very poor.

On the configuration reloading, that is where I tried to reload it from (see screenshot below) but the changed parameters only seem to take effect when restarting Home Assistant itself.

![image](https://github.com/user-attachments/assets/ea1cd9e9-1cc4-4dc5-abd2-26157d24932a)

On those thresholds you mention, the accumulated_error_threshold seems to just be a cap for the integrated error (i.e. the bit multiplied by the ki parameter) applied in the code below:

https://github.com/jmcollin78/versatile_thermostat/blob/60bd522a976caafb1daf30d7d28c8f9fa4db394d/custom_components/versatile_thermostat/pi_algorithm.py#L81

On that stabilization_threshold, the docs seem to imply that the integrated error will be reset when it's within the stabilization_threshold.  However, I'm confused by this, because I can't find anywhere in the code that actually applies this logic or uses this parameter?  Instead there's some logic that reduces it when it crosses through zero:

https://github.com/jmcollin78/versatile_thermostat/blob/60bd522a976caafb1daf30d7d28c8f9fa4db394d/custom_components/versatile_thermostat/pi_algorithm.py#L76

Whilst we're on the subject of the integrated error term, one thing that slightly confused me is that the error integration/accumulation doesn't scale by the time difference?  Surely this means that the more frequently the regulated temperature calculation runs, the faster the integrated error accumulates?   What controls how frequently the calculation runs?

The threshold I was suggesting isn't specific to the integrated error at all.  It's simply a line of logic at the end that says that if the difference between the target temperature and the room temperature is less than the threshold, simply don't update the regulated temperature at all.

**Reply by jmcollin78** [reply]
ok I see. You are refering to the regulation_threshold which designed exactly for this:

"Seuil de régulation" in FR is "Regulation threshold" in EN.
![Capture d’écran 2024-11-02 à 23 45 18](https://github.com/user-attachments/assets/581f6198-e6ba-4f66-8a0f-d82b0d1558ee)

**Reply by t0mb080** [reply]
Ah, I thought that was just the minimum increment between regulated temperature changes.
